pkg/blockbuilder/blockbuilder_test.go

```diff
@@ -1 +1,124 @@
 package blockbuilder
+
+import (
+	"context"
+	"errors"
+	"github.com/go-kit/log"
+	mimir_tsdb "github.com/grafana/mimir/pkg/storage/tsdb"
+	"github.com/grafana/mimir/pkg/util/testkafka"
+	"github.com/grafana/mimir/pkg/util/validation"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"testing"
+	"time"
+)
+
+func TestBlockBuilder_PartitionAssignment(t *testing.T) {
+	// TODO: Implement this test
+	// Test that when there are multiple partitons and block builders, the partitions are assigned properly
+	// without an overlap in partitions and properly balanced. If this makes sense as an integration test,
+	// it can be moved to the integration test file.
+}
+
+func TestBlockBuilder_BuildBlocks(t *testing.T) {
+	// TODO: Implement this test
+	// - Create test kafka cluster.
+	// - add some samples to kafka topic/partitions before starting the blockbuilder.
+	// - start a blockbuilder and test the correct production of blocks on startup.
+	// - then add more samples and test the correct production of blocks after startup.
+	// - Test both float and histogram samples.
+	// - See test cases in tsdb_test.go for inspiration for various cases.
+	// - also test the case of block builder consuming 1 vs more than 1 partition.
+
+	const (
+		topicName   = "test"
+		partitionID = 0
+		userID      = "1"
+	)
+
+	ctx, cancel := context.WithCancelCause(context.Background())
+	t.Cleanup(func() { cancel(errors.New("test done")) })
+
+	_, clusterAddr := testkafka.CreateCluster(t, 1, topicName)
+
+	writeClient := newKafkaProduceClient(t, clusterAddr)
+	for i := int64(0); i < 10; i++ {
+		produceRecord(ctx, t, writeClient, topicName, partitionID, userID, createWriteRequest(t, floatSample(i), nil))
+	}
+
+	cfg := Config{
+		Kafka: KafkaConfig{
+			Address:       clusterAddr,
+			Topic:         topicName,
+			ClientID:      "1",
+			DialTimeout:   10 * time.Second,
+			ConsumerGroup: "testgroup",
+		},
+		BlocksStorageConfig: mimir_tsdb.BlocksStorageConfig{},
+	}
+
+	limits := defaultLimitsTestConfig()
+	limits.OutOfOrderTimeWindow = 2 * model.Duration(time.Hour)
+	limits.NativeHistogramsIngestionEnabled = true
+	overrides, err := validation.NewOverrides(limits, nil)
+
+	bb, err := NewBlockBuilder(cfg, log.NewNopLogger(), nil, overrides)
+	require.NoError(t, err)
+	_ = bb
+
+	require.NoError(t, bb.StartAsync(ctx))
+	require.NoError(t, bb.AwaitRunning(ctx))
+	require.NoError(t, bb.AwaitTerminated(ctx))
+
+	// This works where we dont mention group and do not disable auto commit
+	//opts := []kgo.Opt{
+	//	kgo.SeedBrokers(clusterAddr),
+	//	kgo.ConsumeTopics(topicName),
+	//}
```

Comment:
> This works where we dont mention group and do not disable auto commit

My guess, is that's due to mimir's `testkafka` package, that is explicitly tuned for ingest-store's usage of consumer groups (ref [code][1]). We could start with using `kfake` package directly here instead.

[1]: https://github.com/grafana/mimir/blob/8c632d2dccfb6137b1cf37fa6f8e3c26507f97f6/pkg/util/testkafka/cluster.go#L33-L35